src/Sarif.Multitool/Rules/SARIF1020.SchemaMustBePresentAndConsistent.cs

```diff
@@ -0,0 +1,54 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.CodeAnalysis.Sarif.Multitool.Rules
+{
+    public class SchemaMustBePresentAndConsistent : SarifValidationSkimmerBase
+    {
+        private readonly MultiformatMessageString _fullDescription = new MultiformatMessageString
+        {
+            Text = RuleResources.SARIF1020_SchemaMustBePresentAndConsistent
+        };
+
+        public override MultiformatMessageString FullDescription => _fullDescription;
+
+        public override FailureLevel DefaultLevel => FailureLevel.Error;
+
+        /// <summary>
+        /// SARIF1020
+        /// </summary>
+        public override string Id => RuleId.SchemaMustBePresentAndConsistent;
+
+        protected override IEnumerable<string> MessageResourceNames => new string[]
+        {
+            nameof(RuleResources.SARIF1020_SchemaMustBePresentAndConsistent),
+            nameof(RuleResources.SARIF1020_UseNewSchemaVersion)
+        };
+
+        protected override void Analyze(SarifLog log, string logPointer)
+        {
+            AnalyzeSchema(log.SchemaUri, logPointer);
+        }
+
+        private void AnalyzeSchema(Uri schemaUri, string pointer)
+        {
+            if (schemaUri == null)
+            {
+                LogResult(pointer, nameof(RuleResources.SARIF1020_SchemaMustBePresentAndConsistent));
+                return;
+            }
+
+            if (!schemaUri.OriginalString.EndsWith(VersionConstants.StableSarifVersion)
+                && !schemaUri.OriginalString.EndsWith($"{VersionConstants.StableSarifVersion}.json")
+                && !schemaUri.OriginalString.EndsWith(VersionConstants.SchemaVersionAsPublishedToSchemaStoreOrg)
+                && !schemaUri.OriginalString.EndsWith($"{VersionConstants.SchemaVersionAsPublishedToSchemaStoreOrg}.json"))
+            {
+                LogResult(pointer, nameof(RuleResources.SARIF1020_UseNewSchemaVersion), schemaUri.OriginalString);
```

Comment:
>pointer [](start = 26, length = 7)

The `pointer` argument is a JSON Pointer string like `"/runs/0/results/0"` (which points to the first result in the first run in the log file). In this case, `pointer` will be `/` since we are analyzing the `SarifLog` object which represents the root of the log file.

When you are reporting that the `$schema` property is _missing_, then the location of the error really is `"/"`. But when `$schema` is missing but invalid (because it is old), you actually want to report the error _at the location of the `$schema` property_.

The way you do that is to add a `using` directive for `Microsoft.Json.Pointer`. Then, the first argument to `LogResult` will be `pointer.AtProperty(SarifPropertyNames.Schema)`.

Reply:
I'm going to close this comment and re-open it on the latest iteration, because this comment is not visible in CodeFlow on the latest iteration.

---
In reply to: [433992879](https://github.com/microsoft/sarif-sdk/pull/1892#discussion_r433992879) [](ancestors = 433992879)